build_tools/benchmarks/run_benchmarks_on_android.py

```diff
@@ -183,17 +186,87 @@ def adb_start_cmd(
-        flag_name, flag_value = line.strip().split("=")
-        if flag_name == "--module":
-            # Realpath canonicalization matters. The caller may rely on that to track
-            # which files it already pushed.
-            return (benchmark_case_dir / flag_value).resolve()
-    raise ValueError(f"{flagfile} does not contain a --module flag")
+def adb_path_exists(android_path: pathlib.PurePosixPath, verbose: bool = False):
+    """Run stat to check if the path exists."""
+    proc = adb_start_cmd(["stat", str(android_path)], verbose=verbose)
+    return proc.wait() == 0
+
+
+def adb_fetch_and_push_file(
+    source: Union[str, pathlib.Path],
+    dest: pathlib.PurePosixPath,
+    verbose: bool = False,
+):
+    """Fetch file from the path/URL and stream to the device.
+
+    In the case of fetching, this method avoids the temporary file on the host
+    and reduces the overhead when the file is large.
+
+    Args:
+      source: path/URL to fetch the file.
+      dest: the full dest path on the device.
+      verbose: output verbose message.
+
+    Returns:
+      File path on the device.
+    """
+
+    if adb_path_exists(dest, verbose):
+        return dest
+
+    # If the source is a local file, push directly.
+    if isinstance(source, pathlib.Path):
+        return adb_push_file(source, dest, verbose=verbose)
+
+    if verbose:
+        print(f"Streaming file {source} to {dest}.")
+
+    req = requests.get(source, stream=True, timeout=60)
+    if not req.ok:
+        raise RuntimeError(f"Failed to fetch {source}: {req.status_code} - {req.text}")
+
+    # Implement the ADB sync protocol to stream file chunk to the device, since
+    # the adb client tool doesn't support it.
+    #
+    # Alternatively we can use thrid-party library such as
+    # https://github.com/JeffLIrion/adb_shell. But the protocol we need is
+    # simple and fairly stable. This part can be replaced with other solutions
+    # if needed.
+    #
+    # To understand the details of the protocol, see
+    # https://cs.android.com/android/_/android/platform/packages/modules/adb/+/93c8e3c26e4de3a2b767a2394200bc0721bb1e24:OVERVIEW.TXT
+
+    def wait_ack_ok(sock: socket.socket):
+        buf = bytearray()
+        while len(buf) < 4:
+            data = sock.recv(4 - len(buf))
+            if not data:
+                break
+            buf += data
+
+        if buf.decode("utf-8") != "OKAY":
+            raise RuntimeError(f"ADB communication error: {buf}")
+
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+        sock.connect(ADB_SERVER_ADDR)
+        # Connect to any device (the first 4 hexadecimals is the following text
+        # command length).
+        sock.sendall(b"0012host:transport-any")
+        wait_ack_ok(sock)
+        # Switch to sync mode.
+        sock.sendall(b"0005sync:")
+        wait_ack_ok(sock)
+        # Send the dest file path and file permissions.
+        file_attr = f"{dest},{0o644}".encode("utf-8")
```

Comment:
Curious, what is `0o644` used for?

Reply:
It's the linux file permission: `owner: rw, group: r, other: r`. Added a comment